### PR TITLE
[SPARK-36533][SS] Trigger.AvailableNow for running streaming queries like Trigger.Once in multiple batches

### PR DESCRIPTION
**Pull request by bozhang2820**
### What changes were proposed in this pull request?

This change creates a new type of Trigger: Trigger.AvailableNow for streaming queries. It is like Trigger.Once, which process all available data then stop the query, but with better scalability since data can be processed in multiple batches instead of one.

To achieve this, this change proposes a new interface `SupportsTriggerAvailableNow`, which is an extension of `SupportsAdmissionControl`. It has one method, `prepareForTriggerAvailableNow`, which will be called at the beginning of streaming queries with Trigger.AvailableNow, to let the source record the offset for the current latest data at the time (a.k.a. the target offset for the query). The source should then behave as if there is no new data coming in after the beginning of the query, i.e., the source will not return an offset higher than the target offset when `latestOffset` is called.

This change also updates `FileStreamSource` to be an implementation of `SupportsTriggerAvailableNow`.

For other sources that does not implement `SupportsTriggerAvailableNow`, this change also provides a new class `FakeLatestOffsetSupportsTriggerAvailableNow`, which wraps the sources and makes them support Trigger.AvailableNow, by overriding their `latestOffset` method to always return the latest offset at the beginning of the query.

### Why are the changes needed?

Currently streaming queries with Trigger.Once will always load all of the available data in a single batch. Because of this, the amount of data a query can process is limited, or Spark driver will run out of memory. 

### Does this PR introduce _any_ user-facing change?

Users will be able to use Trigger.AvailableNow (to process all available data then stop the streaming query) with this change.

### How was this patch tested?

Added unit tests.

### CODE REVIEW

**Comment by bozhang2820** [comment]
@HeartSaVioR, @brkyvz, could you review this? Thanks!

**Comment by xuanyuanking** [comment]
cc @viirya

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:56]: It'd be better to have consistent name rule between trigger and executor. `MultiBatchExecutor` doesn't represent the semantic from its name.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TriggerAvailableNowSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TriggerAvailableNowSuite.scala:144]: (Beyond the scope) it would be ideal if we can revisit and fix it.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TriggerAvailableNowSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TriggerAvailableNowSuite.scala:99]: Let's make the test suite be exhaustive; 1) single source which doesn't support Trigger.AvailableNow 2) single source which supports Trigger.AvailableNow.

Latter can be covered by new FileStreamSource test, hence adding test for 1) would be enough.

**Review by HeartSaVioR — COMMENTED** [review]
First pass.

**Comment by HeartSaVioR** [comment]
Btw, thanks for the great contribution! Nice feature indeed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetMicroBatchStream.scala
Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetMicroBatchStream.scala:26]: Will do.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala
Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala:52]: Thanks for catching this. Will do.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala
Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:56]: `MultiBatchExecutor` is really just running `batchRunner` multiple times. The semantic of processing all available data is implemented in `MicroBatchExecution`.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TriggerAvailableNowSuite.scala
Comment (bozhang2820) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/TriggerAvailableNowSuite.scala:99]: Sure. Will do.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:56]: Ah OK. I agree that the executor can be separated from the trigger and be generalized by itself.

**Review by xuanyuanking — COMMENTED** [review]
(Seems I failed to post the review comments last 4 days ago...)
Still reviewing this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala
Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala:52]: I'm getting:

```
[error] Error while emitting FileStreamSource.scala
[error] assertion failed: 
[error]   cannot invokespecial org/apache/spark/sql/connector/read/streaming/SupportsAdmissionControl.getDefaultReadLimit, the interface is not a direct parent.
```

after removing `SupportsAdmissionControl`. Can I keep it and add a TODO here for now?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetMicroBatchStream.scala
Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetMicroBatchStream.scala:29]: Done.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetMicroBatchStream.scala
Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetMicroBatchStream.scala]: Done.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala
Comment (bozhang2820) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:417]: We can also combine this with the following pattern, but that would need to change the type of `originalSource` to be more generic (`SupportsAdmissionControl` instead of `FakeLatestOffsetSupportsTriggerAvailableNow`).

**Review by viirya — COMMENTED** [review]
Thanks for the work. I will review in next few days.

I think we also need to update the document. But I don't see doc change included yet.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/streaming/SupportsTriggerAvailableNow.java
Comment (viirya) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/streaming/SupportsTriggerAvailableNow.java:28]: Could you add `@since` annotation too? E.g. `SupportsAdmissionControl`.

File: sql/core/src/main/java/org/apache/spark/sql/streaming/Trigger.java
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/streaming/Trigger.java:93]: Maybe add "For better scalability,  `AvailableNow` can be used alternatively to process it in multiple batches.".

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Ah got it. It's OK to not touch if it breaks the compilation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala:52]: Let's remove the TODO and never mind if it fails the compilation.

**Review by HeartSaVioR — COMMENTED** [review]
Looks OK in overall. I'd like to wait for @viirya 's review and approval before merging.

**Comment by bozhang2820** [comment]
> Thanks for the work. I will review in next few days.
> 
> I think we also need to update the document. But I don't see doc change included yet.

Thanks @viirya! Could you point me to some places where I need to update the doc?

**Comment by viirya** [comment]
The structured streaming programming guide: https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html#triggers

The source is at docs/structured-streaming-programming-guide.md.

File: sql/core/src/main/java/org/apache/spark/sql/streaming/Trigger.java
Comment (brkyvz) [inline sql/core/src/main/java/org/apache/spark/sql/streaming/Trigger.java:23]: uber nit (also existing): isn't the import order messed up? Shouldn't scala be it's own group. the it should be 
```java
import org.apache.spark.annotation.Evolving;
import org.apache.spark.sql.execution.streaming.AvailableNowTrigger$;
import org.apache.spark.sql.execution.streaming.ContinuousTrigger;import 
import org.apache.spark.sql.execution.streaming.OneTimeTrigger$;
import org.apache.spark.sql.execution.streaming.ProcessingTimeTrigger;

```

File: sql/core/src/main/java/org/apache/spark/sql/streaming/Trigger.java
Comment (brkyvz) [inline sql/core/src/main/java/org/apache/spark/sql/streaming/Trigger.java:102]: `... all available data at the start of the query in one or multiple batches...`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetMicroBatchStream.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetMicroBatchStream.scala:32]: should this be throwing an UnsupportedOperationException? I was confused on why you'd need to override this, and later realized that this wasn't the rate limited version

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetSource.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetSource.scala:33]: ditto - if this method isn't going to be called, let's have it throw an UnsupportedOperationException

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:56]: uber nit: `SingleBatchExecutor`, while is longer better fits with `MultiBatchExecutor`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:417]: As you actually pointed out here, I no longer see the value of having the `originalSources` map. Am I missing anything? Can we just do the unwrapping transparently inside `StreamProgress` for `FakeLatestOffsetSupportsTriggerAvailableNow`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetSupportsTriggerAvailableNow.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetSupportsTriggerAvailableNow.scala:30]: can we call this `AvailableNowSourceWrapper(val delegate: SparkDataStream)`? Then you can special case this in `StreamProgress` and not need `originalSources` and things like that. You can even override the `toString` method to return the `delegate.toString` so that the metrics look correct as well

**Review by brkyvz — CHANGES_REQUESTED** [review]
This looks awesome! Left some comments. It would also be great to add tests around the progress reporting (not overwriting the Source description with that of the wrapper class, etc)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala
Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:417]: For that we would have to wrap sources in the logical plan... I think having the wrapped sources only in `MicroBatchExecution.uniqueSources` is a less intrusive and safer change. What do you think?

I will also remove `originalSource` mapping and get the original source directly from the wrapped source instead.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:417]: If we use `s.source` here to extract the original source, then we can get rid of `originalSource` as well since this is the only usage.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetSupportsTriggerAvailableNow.scala
Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetSupportsTriggerAvailableNow.scala:30]: I can do the rename here. For the places to wrap the sources let's have the discussion in the other thread.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala
Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:417]: Thanks @HeartSaVioR! Will follow what you suggest and remove `originalSource`.

**Comment by bozhang2820** [comment]
> This looks awesome! Left some comments. It would also be great to add tests around the progress reporting (not overwriting the Source description with that of the wrapper class, etc)

Thanks! Added assertions for the source descriptions in `TriggerAvailableNowSuite`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala:30]: I guess what @brkyvz said is to override `toString` here to delegate to the origin `source`.

Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala:30]: I think that is not needed because we are still using the original source names in `ProgressReporter`. Let me know if you think otherwise.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala:30]: You're right. We only wrap sources in uniqueSources which won't affect ProgressReporter.

**Review by HeartSaVioR — DISMISSED** [review]
+1

Wrapping sources makes lookup be complicated, but it's limited to the `uniqueSources` which seems to be maintainable IMHO.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala:43]: As this implements `SupportsTriggerAvailableNow`, but seems `prepareForTriggerAvailableNow` is useless? Why don't we record latest offset here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala:112]: `private`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala:134]: Should this comment follow `fetchAllFiles`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:146]: Should we log warning if it is `getDefaultReadLimit` is `ReadLimit.allAvailable()` for `AvailableNowTrigger`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:412]: here shouldn't we match against `originalSource`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:266]: all data -> all available data for the trigger

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala
Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala:134]: Will update the comment here to describe the logic for Trigger.AvailableNow instead.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala
Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala:43]: Thanks for the suggestion! This made me realized that we might need to have `startOffset` as an argument in `prepareForTriggerAvailableNow`

Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala]: Turned out `startOffset` is not needed in `prepareForTriggerAvailableNow`. See the latest code update.

**Comment by xuanyuanking** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala
Comment (bozhang2820) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala:46]: We can use null (for v1 sources) and iniitalOffset (for v2 streams) for the startOffset in latestOffset(startOffset, readLimit), since the readLimit is always allAvailable.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
@viirya Would you mind revisiting this? The PR looks OK to me, but I'd like to see your approval as you're reviewing the change. Thanks in advance!

**Comment by bozhang2820** [comment]
> @viirya Would you mind revisiting this? The PR looks OK to me, but I'd like to see your approval as you're reviewing the change. Thanks in advance!

Also @brkyvz could you review again for the changes you requested?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala:1340]: Shouldn't we add more offsets `createFile` here after latest offset was fetched into the source? So we can verify that we only process all available at the beginning of the query?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TriggerAvailableNowSuite.scala
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TriggerAvailableNowSuite.scala:206]: ditto. Should we increase more offsets here after triggering the query?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala
Comment (bozhang2820) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala:1340]: I find it a bit hard to control the order between starting the query and adding new files into the source. Do you know if there is an easy way to do so?

Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala:1340]: I'm not sure we could do it easily, as we can't let the query be "suspended" after figuring out source offsets to process. The streaming query is running concurrently with the main thread.

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala:1340]: Hmm, okay, I asked this because the test looks like to me only for verifying the query can run with correct results. Not able to check trigger.availablenow. okay for me.

Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala:1340]: `assert(q.recentProgress.count(_.numInputRows != 0) == 3)`

This verifies that the query runs three micro-batches instead of one.

Btw, the test `SPARK-36533: Trigger.AvailableNow - checkpointing` covers everything in here now. The changes on `index` determines the number of micro-batch being executed, and we also checked the output DataFrame. I think we can simply remove this test.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala:1403]: nit: probably better to add the comment that it tracks the number of micro-batch execution starting from here. The code is intuitive but worth having elaboration given the importance of this variable.

**Review by HeartSaVioR — APPROVED** [review]
+1 pending other reviewers' approval.

**Review by viirya — APPROVED** [review]

**Review by brkyvz — APPROVED** [review]
LGTM!

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master!

**Comment by HeartSaVioR** [comment]
Thanks @bozhang2820 for the great contribution and thanks all for reviewing! I just merged this into master.

**Comment by srowen** [comment]
@HeartSaVioR @viirya @bozhang2820 is there a way to call this from Pyspark? I can't figure it out

**Comment by HeartSaVioR** [comment]
Nice catch. I realized it was missing. I'm going to address this soon.

**Pull request closed without merging**
